src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,237 @@
+instance : has_one (centroid_hom α) := ⟨centroid_hom.id α⟩
+
+instance : has_add (centroid_hom α) :=
+⟨λ f g, ⟨f + g, λ a b, by simp [map_mul_left, mul_add], λ a b, by simp [map_mul_right, add_mul]⟩⟩
+
+instance : has_mul (centroid_hom α) := ⟨comp⟩
+
+instance has_scalar_nat : has_scalar ℕ (centroid_hom α) :=
+⟨λ n f, ⟨n • f,
+  λ a b, by { change n • f (a * b) = a * n • f b, rw [map_mul_left f, mul_nsmul_left] },
+  λ a b, by { change n • f (a * b) = n • f a * b, rw [map_mul_right f, mul_nsmul_assoc] }⟩⟩
+
+instance has_npow_nat : has_pow (centroid_hom α) ℕ :=
+⟨λ f n, ⟨(f.to_End ^ n : add_monoid.End α), λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_left' _ _) }
+  end, λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_right' _ _) }
+  end⟩⟩
+
+@[simp] lemma coe_zero : ⇑(0 : centroid_hom α) = 0 := rfl
+@[simp] lemma coe_one : ⇑(1 : centroid_hom α) = id := rfl
+@[simp] lemma coe_add (f g : centroid_hom α) : ⇑(f + g) = f + g := rfl
+@[simp] lemma coe_mul (f g : centroid_hom α) : ⇑(f * g) = f ∘ g := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma coe_nsmul (f : centroid_hom α) (n : ℕ) : ⇑(n • f) = n • f := rfl
+
+@[simp] lemma zero_apply (a : α) : (0 : centroid_hom α) a = 0 := rfl
+@[simp] lemma one_apply (a : α) : (1 : centroid_hom α) a = a := rfl
+@[simp] lemma add_apply (f g : centroid_hom α) (a : α) : (f + g) a = f a + g a := rfl
+@[simp] lemma mul_apply (f g : centroid_hom α) (a : α) : (f * g) a = f (g a) := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf]
+lemma nsmul_apply (f : centroid_hom α) (n : ℕ) (a : α) : (n • f) a = n • f a := rfl
+
+-- cf.`add_monoid_hom.add_comm_monoid`
+instance : add_comm_monoid (centroid_hom α) :=
+coe_to_add_monoid_hom_injective.add_comm_monoid _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+-- cf `add_monoid.End.semiring`
+instance : semiring (centroid_hom α) :=
+to_End_injective.semiring _ (by { ext, refl }) (by { ext, refl }) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl) (λ f a, by { ext, refl })
+
+lemma comp_mul_comm (T S : centroid_hom α) (a b : α) : (T ∘ S) (a * b) = (S ∘ T) (a * b) :=
+by rw [comp_app, map_mul_right, map_mul_left, ←map_mul_right, ←map_mul_left]
+
+end non_unital_non_assoc_semiring
+
+section non_unital_non_assoc_ring
+variables [non_unital_non_assoc_ring α]
+
+/-- Negation of `centroid_hom`s as a `centroid_hom`. -/
+instance : has_neg (centroid_hom α) :=
+⟨λ f, ⟨-f, by simp [map_mul_left], by simp [map_mul_right]⟩⟩
+
+instance : has_sub (centroid_hom α) :=
+⟨λ f g, ⟨f - g, λ a b, by simp [map_mul_left, mul_sub], λ a b, by simp [map_mul_right, sub_mul]⟩⟩
+
+instance has_scalar_int : has_scalar ℤ (centroid_hom α) :=
+⟨λ n f, ⟨n • f,
+  λ a b, by { change n • f (a * b) = a * n • f b, rw [map_mul_left f, mul_zsmul_left] },
+  λ a b, by { change n • f (a * b) = n • f a * b, rw [map_mul_right f, mul_zsmul_assoc] }⟩⟩
```

Comment:
I think there should be a more general version of this, see https://leanprover-community.github.io/mathlib_docs/algebra/module/hom.html for the type of thing I'm thinking of.